components/script/script_module.rs

```diff
@@ -1864,14 +1867,142 @@ pub(crate) fn fetch_inline_module_script(
     }
 }
 
-#[derive(Default, JSTraceable)]
+pub(crate) type ModuleSpecifierMap = IndexMap<String, Option<ServoUrl>>;
+pub(crate) type ModuleIntegrityMap = IndexMap<ServoUrl, String>;
+
+/// <https://html.spec.whatwg.org/multipage/#import-map-processing-model>
+#[derive(Default, JSTraceable, MallocSizeOf)]
 pub(crate) struct ImportMap {
     #[no_trace]
-    imports: IndexMap<String, Option<ServoUrl>>,
+    imports: ModuleSpecifierMap,
     #[no_trace]
-    scopes: IndexMap<String, IndexMap<String, Option<ServoUrl>>>,
+    scopes: IndexMap<ServoUrl, ModuleSpecifierMap>,
     #[no_trace]
-    integrity: IndexMap<String, String>,
+    integrity: ModuleIntegrityMap,
+}
+
+/// <https://html.spec.whatwg.org/multipage/#register-an-import-map>
+pub(crate) fn register_import_map(
+    global: &GlobalScope,
+    result: &Fallible<ImportMap>,
+    can_gc: CanGc,
+) {
+    match result {
+        Ok(new_import_map) => {
+            // Step 2. Merge existing and new import maps, given global and result's import map.
+            merge_existing_and_new_import_maps(global, new_import_map, can_gc);
+        },
+        Err(exception) => {
+            // Step 1. If result's error to rethrow is not null, then report
+            // an exception given by result's error to rethrow for global and return.
+            throw_dom_exception(GlobalScope::get_cx(), global, exception.clone(), can_gc);
+        },
+    }
+}
+
+/// <https://html.spec.whatwg.org/multipage/#merge-existing-and-new-import-maps>
+fn merge_existing_and_new_import_maps(
+    global: &GlobalScope,
+    new_import_map: &ImportMap,
+    can_gc: CanGc,
+) {
+    // Step 1. Let newImportMapScopes be a deep copy of newImportMap's scopes.
+    let new_import_map_scopes = new_import_map.scopes.clone();
+
+    // Step 2. Let oldImportMap be global's import map.
+    let mut old_import_map = global.import_map().borrow_mut();
+
+    // Step 3. Let newImportMapImports be a deep copy of newImportMap's imports.
+    let new_import_map_imports = new_import_map.imports.clone();
```

Comment:
Deep copies here could be obsolete. We only need to clone url in `merge_module_specifier_maps`.